File: app/src/main/java/org/thoughtcrime/securesms/recipients/ui/sharablegrouplink/ShareableGroupLinkRepository.java
package org.thoughtcrime.securesms.recipients.ui.sharablegrouplink;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;

import org.signal.core.util.concurrent.SignalExecutors;
import org.signal.core.util.logging.Log;
import org.signal.storageservice.storage.protos.groups.AccessControl;
import org.thoughtcrime.securesms.database.SignalDatabase;
import org.thoughtcrime.securesms.database.model.GroupRecord;
import org.thoughtcrime.securesms.groups.GroupChangeBusyException;
import org.thoughtcrime.securesms.groups.GroupChangeFailedException;
import org.thoughtcrime.securesms.groups.GroupId;
import org.thoughtcrime.securesms.groups.GroupInsufficientRightsException;
import org.thoughtcrime.securesms.groups.GroupManager;
import org.thoughtcrime.securesms.groups.GroupNotAMemberException;
import org.thoughtcrime.securesms.groups.ui.GroupChangeFailureReason;
import org.thoughtcrime.securesms.util.AsynchronousCallback;

import java.io.IOException;

final class ShareableGroupLinkRepository {

  private static final String TAG = Log.tag(ShareableGroupLinkRepository.class);

  private final Context    context;
  private final GroupId.V2 groupId;

  ShareableGroupLinkRepository(@NonNull Context context, @NonNull GroupId.V2 groupId) {
    this.context = context;
    this.groupId = groupId;
  }

  void cycleGroupLinkPassword(@NonNull AsynchronousCallback.WorkerThread<Void, GroupChangeFailureReason> callback) {
    SignalExecutors.UNBOUNDED.execute(() -> {
      try {
        GroupManager.cycleGroupLinkPassword(context, groupId);
        callback.onComplete(null);
      } catch (GroupNotAMemberException | GroupChangeFailedException | GroupInsufficientRightsException | IOException | GroupChangeBusyException e) {
        callback.onError(GroupChangeFailureReason.fromException(e));
      }
    });
  }

  void toggleGroupLinkEnabled(@NonNull AsynchronousCallback.WorkerThread<Void, GroupChangeFailureReason> callback) {
    setGroupLinkEnabledState(true, false, callback);
  }

  void toggleGroupLinkApprovalRequired(@NonNull AsynchronousCallback.WorkerThread<Void, GroupChangeFailureReason> callback) {
    setGroupLinkEnabledState(false, true, callback);
  }

  private void setGroupLinkEnabledState(boolean toggleEnabled,
                                        boolean toggleApprovalNeeded,
                                        @NonNull AsynchronousCallback.WorkerThread<Void, GroupChangeFailureReason> callback)
  {
    SignalExecutors.UNBOUNDED.execute(() -> {
      GroupRecord groupRecord = SignalDatabase.groups().getGroup(groupId).orElse(null);

      if (groupRecord == null || !groupRecord.getHasV2GroupProperties()) {
        Log.w(TAG, "Unable to find group, likely deleted.");
        callback.onError(GroupChangeFailureReason.OTHER);
        return;
      }

      try {
        GroupManager.setGroupLinkEnabledState(context, groupId, toggleGroupLinkState(groupRecord, toggleEnabled, toggleApprovalNeeded));
        callback.onComplete(null);
      } catch (GroupNotAMemberException | GroupChangeFailedException | GroupInsufficientRightsException | IOException | GroupChangeBusyException e) {
        callback.onError(GroupChangeFailureReason.fromException(e));
      }
    });
  }

  @WorkerThread
  private GroupManager.GroupLinkState toggleGroupLinkState(@NonNull GroupRecord groupRecord, boolean toggleEnabled, boolean toggleApprovalNeeded) {
    //noinspection DataFlowIssue
    AccessControl.AccessRequired currentState = groupRecord.requireV2GroupProperties()
                                                           .getDecryptedGroup()
                                                           .accessControl
                                                           .addFromInviteLink;

    boolean enabled;
    boolean approvalNeeded;

    switch (currentState) {
      case UNKNOWN:
      case UNSATISFIABLE:
      case MEMBER:
        enabled        = false;
        approvalNeeded = false;
        break;
      case ANY:
        enabled        = true;
        approvalNeeded = false;
        break;
      case ADMINISTRATOR:
        enabled        = true;
        approvalNeeded = true;
        break;
      default: throw new AssertionError();
    }

    if (toggleApprovalNeeded) {
      approvalNeeded = !approvalNeeded;
    }

    if (toggleEnabled) {
      enabled = !enabled;
    }

    if (approvalNeeded && enabled) {
      return GroupManager.GroupLinkState.ENABLED_WITH_APPROVAL;
    } else {
      if (enabled) {
        return GroupManager.GroupLinkState.ENABLED;
      }
    }

    return GroupManager.GroupLinkState.DISABLED;
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/recipients/ui/sharablegrouplink/ShareableGroupLinkRepository.kt
/*
 * Copyright 2026 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.thoughtcrime.securesms.recipients.ui.sharablegrouplink

import android.content.Context
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.rx3.asFlow
import kotlinx.coroutines.withContext
import org.signal.chatsettings.screens.sharablegrouplink.GroupLink
import org.signal.core.util.concurrent.SignalDispatchers
import org.signal.core.util.groups.GroupChangeException
import org.signal.core.util.logging.Log
import org.signal.core.util.orNull
import org.signal.storageservice.storage.protos.groups.AccessControl
import org.thoughtcrime.securesms.database.GroupTable
import org.thoughtcrime.securesms.database.SignalDatabase
import org.thoughtcrime.securesms.database.model.GroupRecord
import org.thoughtcrime.securesms.dependencies.AppDependencies
import org.thoughtcrime.securesms.groups.GroupId
import org.thoughtcrime.securesms.groups.GroupManager
import org.thoughtcrime.securesms.groups.ui.GroupChangeFailureReason
import org.thoughtcrime.securesms.groups.ui.GroupChangeResult
import org.thoughtcrime.securesms.groups.v2.GroupInviteLinkUrl
import org.thoughtcrime.securesms.recipients.Recipient
import java.io.IOException

/**
 * All of the group link reads and writes behind [ShareableGroupLinkViewModel].
 */
class ShareableGroupLinkRepository(
  private val context: Context = AppDependencies.application,
  private val groupTable: GroupTable = SignalDatabase.groups
) {

  companion object {
    private val TAG = Log.tag(ShareableGroupLinkRepository::class)
  }

  /**
   * Emits the group's link whenever it changes. A group change always touches the group's recipient, so recipient
   * updates drive this. Plenty of unrelated activity touches it too -- an incoming message refreshes it -- hence the
   * dedupe.
   */
  fun observeGroupLink(groupId: GroupId.V2): Flow<GroupLink> {
    return flow {
      val recipientId = withContext(SignalDispatchers.Default) { Recipient.externalGroupExact(groupId).id }
      emitAll(Recipient.observable(recipientId).asFlow())
    }.mapNotNull { recipient ->
      withContext(SignalDispatchers.Default) {
        groupTable.getGroup(recipient.id).orNull()?.toGroupLink()
      }
    }.distinctUntilChanged()
  }

  /**
   * Turns the link on or off, and sets whether joining through it needs an admin's approval.
   */
  suspend fun setGroupLinkState(groupId: GroupId.V2, enabled: Boolean, requiresAdminApproval: Boolean): GroupChangeResult = groupChange {
    val state = when {
      !enabled -> GroupManager.GroupLinkState.DISABLED
      requiresAdminApproval -> GroupManager.GroupLinkState.ENABLED_WITH_APPROVAL
      else -> GroupManager.GroupLinkState.ENABLED
    }

    GroupManager.setGroupLinkEnabledState(context, groupId, state)
  }

  /**
   * Gives the group a new link, which stops the previous one from working.
   */
  suspend fun cycleGroupLinkPassword(groupId: GroupId.V2): GroupChangeResult = groupChange {
    GroupManager.cycleGroupLinkPassword(context, groupId)
  }

  /**
   * Runs a group change, which hits the network, and translates whatever went wrong into something we can show the
   * user.
   */
  private suspend fun groupChange(block: () -> Unit): GroupChangeResult {
    return withContext(SignalDispatchers.IO) {
      try {
        block()
        GroupChangeResult.SUCCESS
      } catch (e: GroupChangeException) {
        Log.w(TAG, e)
        GroupChangeResult.failure(GroupChangeFailureReason.fromException(e))
      } catch (e: IOException) {
        Log.w(TAG, e)
        GroupChangeResult.failure(GroupChangeFailureReason.fromException(e))
      }
    }
  }

  private fun GroupRecord.toGroupLink(): GroupLink {
    val selfCanEditSettings = isActive && isAdmin(Recipient.self())

    if (!hasV2GroupProperties) {
      return GroupLink.NONE.copy(selfCanEditSettings = selfCanEditSettings)
    }

    val properties = requireV2GroupProperties()
    val decryptedGroup = properties.decryptedGroup

    if (decryptedGroup.inviteLinkPassword.size == 0) {
      return GroupLink.NONE.copy(selfCanEditSettings = selfCanEditSettings)
    }

    return GroupLink(
      enabled = isGroupLinkEnabled,
      requiresAdminApproval = decryptedGroup.accessControl?.addFromInviteLink == AccessControl.AccessRequired.ADMINISTRATOR,
      url = GroupInviteLinkUrl.forGroup(properties.groupMasterKey, decryptedGroup).url,
      selfCanEditSettings = selfCanEditSettings
    )
  }
}
